packages/protobuf/src/equals.ts

```diff
@@ -135,3 +193,91 @@ function fieldEquals(
   }
   return true;
 }
+
+function anyUnpackedEquals(a: Any, b: Any, opts: EqualsOptions): boolean {
+  if (a.typeUrl !== b.typeUrl) {
+    return false;
+  }
+  const unpackedA = anyUnpack(a, opts.registry);
+  const unpackedB = anyUnpack(b, opts.registry);
+  if (unpackedA && unpackedB) {
+    const schema = opts.registry.getMessage(unpackedA.$typeName);
+    if (schema) {
+      return equals(schema, unpackedA, unpackedB, opts);
+    }
+  }
+  return scalarEquals(ScalarType.BYTES, a.value, b.value);
+}
+
+function unknownEquals(
+  a: ReflectMessage,
+  b: ReflectMessage,
+  registry: Registry | undefined,
+) {
+  function getTrulyUnknown(
+    msg: ReflectMessage,
+    registry: Registry | undefined,
+  ) {
+    const u = msg.getUnknown() ?? [];
+    return registry
+      ? u.filter((uf) => !registry.getExtensionFor(msg.desc, uf.no))
+      : u;
+  }
+  const unknownA = getTrulyUnknown(a, registry);
+  const unknownB = getTrulyUnknown(b, registry);
+  if (unknownA.length != unknownB.length) {
+    return false;
+  }
+  for (let i = 0; i < unknownA.length; i++) {
+    const a = unknownA[i],
+      b = unknownB[i];
+    if (a.no != b.no) {
+      return false;
+    }
+    if (a.wireType != b.wireType) {
+      return false;
+    }
+    if (!scalarEquals(ScalarType.BYTES, a.data, b.data)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function extensionsEquals(
+  a: ReflectMessage,
+  b: ReflectMessage,
+  opts: EqualsOptions,
+) {
+  function getSetExtensions(
+    msg: ReflectMessage,
+    registry: Registry,
+  ): DescExtension[] {
+    return (msg.getUnknown() ?? [])
+      .map((uf) => registry.getExtensionFor(msg.desc, uf.no))
+      .filter((e) => !!e)
```

Comment:
Is this just removing unknown values? If so, a little surprising that this is idiomatic instead of something more intuitive/readable 🤷.

Reply:
Yes, it's removing unknown values. We can make this a bit more readable - with https://github.com/bufbuild/protobuf-es/pull/1029/commits/410c9393a220f7be6e1dd6c9194fb1d486f61f1d, it's `e != undefined` now.

It's actually a new feature that the TypeScript compiler can narrow down the expression type from `filter` calls. I was confused that it worked for `!!e`, but didn't appear to work for `e != undefined`, but it actually does.